src/providers/postgres/qgspostgresfeatureiterator.cpp

```diff
@@ -41,6 +41,21 @@ QgsPostgresFeatureIterator::QgsPostgresFeatureIterator( QgsPostgresFeatureSource
   if ( !source->mTransactionConnection )
   {
     mConn = QgsPostgresConnPool::instance()->acquireConnection( mSource->mConnInfo, request.timeout(), request.requestMayBeNested() );
+
+    const QString sessionRoleKey = QStringLiteral( "session_role" );
+    if ( mSource->mUri.hasParam( sessionRoleKey ) )
+    {
+      const QString sessionRole = mSource->mUri.param( sessionRoleKey );
+      if ( !sessionRole.isEmpty() )
+      {
+        mConn->setSessionRole( sessionRole );
+      }
+    }
+    else
+    {
+      mConn->resetSessionRole();
```

Comment:
Don't we get unnecessary PostgreSQL request hiere, which #53426 tries to avoid?

Reply:
We do sometimes call reset unnecessarily, but we don't know if that particular connection has been used before and has an undesired role.

We could maybe store the current role in the connection itself and check if we need to change it.

Reply:
The `setSessionRole` and `resetSessionRole` check what is the current role and only update if the requested role is different. So we shouldn't get any unnecessary calls.